src/main/scala/is/hail/utils/BufferedAggregatorIterator.scala

```diff
@@ -0,0 +1,61 @@
+package is.hail.utils
+
+import java.util
+
+import is.hail.annotations.aggregators.{RegionValueAggregator, RegionValueCountAggregator}
+
+class BufferedAggregatorIterator[T, U, K](
+  it: Iterator[T],
+  makeComb: () => U,
+  makeKey: T => K,
+  sequence: (T, U) => Unit,
+  bufferSize: Int
+) extends Iterator[(K, U)] {
+
+  private val buffer = new Array[(K, U)](bufferSize)
+  private val keyIdx: util.HashMap[K, Int] = new util.HashMap()
+  private var ptr = 0
+
+  private def bump(idx: Int): Unit = {
+    if (idx > 0) {
+      val curr = buffer(idx)
+      val prev = buffer(idx - 1)
+      keyIdx.put(curr._1, idx - 1)
+      keyIdx.put(prev._1, idx)
+      buffer(idx) = prev
+      buffer(idx - 1) = curr
+    }
+  }
+
+  private def fill() {
+    while (ptr < bufferSize && it.hasNext) {
```

Comment:
The aggregator for the next element might still reside in the buffer.

The keys should be an LRU cache.  This can be implemented trivially with Java's LinkedHashMap, see @jigold's cache code in FASTAReader.  Then you want `next` to be something like:

```
  while (it.hasNext) {
    if (key in buffer)
      add to aggregator
    else {
       remove oldest
       add new row to buffer
       return oldest
     }
     pop element from buffer
  }
```

Of course, you'll either need to implement a Flipbook iterator or do that in hasNext and store the next in a member variable or something.

Reply:
@patrick-schultz also had some ideas re: heavy hitters algorithms, but this also seems very easy

Reply:
there are probably key distributions where each algorithm is better, though

Reply:
Yeah, the state-of-the-art heavy hitters algorithms that work in the regime this needs (the *turnstile* model, no removals/negative weights) are extremely simple and efficient (possibly more efficent, as they don't need the linked list). I've convinced myself that solutions to the map-side combine problem are exactly solutions to the standard (time-decayed) heavy-hitters problem: map-side combine wants to maximize the number of combine operations it does, which is equivalent to maximizing the probability that each key it sees is already in its cache, so the cache should maintain the most frequent keys seen recently.

The LRU cache's solution is to estimate the n most frequent keys in the recent past to be the n most recently seen, which is probably not a bad approximation for many distributions. It would be interesting to do a benchmark comparing the LRU cache with a few heavy-hitters algorithms (including time-decaying weights with different decay speeds). But for now, I'm fine with getting any one of the reasonable approaches in.

Reply:
added LRU.